### ISSUE TITLE
should in parameters be const?

### ISSUE THREAD

**Issue by shaneasd** [hsutter/708#20]
Equivalently: if the argument is a non-const rvalue then why should _only the definite last use_ of x that would be legal for an rvalue treat x as a non-const rvalue?
Equivalently:  by value non-const parameters seem to have been removed. Is that intentional and/or desriable?

It seems that with, the exception of last use, any parameter which is not used to return information cannot be modified. Again with that exception, that would be the equivalent of defining all my functions as ```foo(const MyType x)```. I think in many cases this is probably a good thing. I typically leave parameters as non-const because I don't want my caller to have to care whether I'm going to modify by-value arguments internally but I also would like to make them const because generally things should be const and it's pretty hard to name a parameter whose meaning changes during the method.

However I have definitely encountered code where a by-value parameter is modified. I think often this is because of laziness but I wonder if there could be an argument for efficiency. I suspect this case would be rare but it would be good for this new system to be at least as powerful as the current system to avoid having to use both.

One example might be if I have this code
```
void PropagateModified(in MyType x)
{
    x.y = 4;
    Propagate(x);
}
```
I'm passing in a construct containing all the data I need but I will have to replace some of it.

Currently my understanding is that this wouldn't compile because x.y is not the definite last use of x and so x is const. The closest equivalent I can see is
```
void PropagateModified(in MyType x)
{
    MyType x2 = x;
    x2.y = 4;
    Propagate(std::move(x2));
}
```

However it seems to me that if the argument is a definite last use or xvalue (or whatever exotic Nvalue categories I'm overlooking) then the creation of x2 is redundant. I think this would at least require one superfluous move construction (compared to the ideal).

It seems to me that ```in``` is an agreement between the caller and callee and the caller doesn't care what the callee does with an argument in the && case.

Presumably the reason to make it const is to permit const& arguments which makes sense but there's already a special case for if the argument happens to be a non-const rvalue. It's a shame that special case must exist but, given that it does, I can't see why it should be limited to the last use.

In short it would be nice to see the mutating parameter use case addressed even if just by saying "it's a bad/obscure idea. Don't do it".

**Comment by shaneasd** [comment]
I note that this is discussed partially (specifically the option to copy into a local variable) in point 3) in the notes section on in parameters.

**Comment by hsutter** [comment]
That's right, `in` parameters are implicitly `const`, they are "a parameter that the function body can read from" and "treated as a `const` lvalue" in the function body.

Re this:
> Equivalently: if the argument is a non-const rvalue then why should only the definite last use of x that would be legal for an rvalue treat x as a non-const rvalue?

I think you're suggesting that if the argument is non-`const` then the function body should be able to treat it as non-`const`. But that doesn't work with static compilation... we need to be able to separately compile the function body once, doing overload resolution etc. as we go, and know that it's correct for all uses. And that use is `const`, because an `in` parameter expresses that we want "an `X` we can read from." As it is, we make a definite last use essentially a "dynamic" call that can invoke a non-`const` overload, but that's a special case; doing it for all uses would be hard to understand and not follow the existing practice of "pass in parameters by `const&`, and if you need to optimize for rvalues then overload on `&&` and `move` from the last use."

**Comment by shaneasd** [comment]
So in effect the difference is that Propagate(x) must be valid for both const& and && whereas ```x.y = 4;``` is clearly not valid for &&. Makes sense.
